Iot Hub Commandlets

rkmanda opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
<!--
Please add an informative description that covers the changes made by the pull request.

If applicable, reference the bug/issue that this pull request fixes here.
-->

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [ ] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThrough` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

azurecla commented:
Hi __@rkmanda__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (rmanda). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

rkmanda commented:
This is the next iteration after the older PR for the same was closed. Link to PR https://github.com/Azure/azure-powershell/pull/3156

markcowl commented:
@rkmanda Please clean up the commits in this PR as described here:https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md

Please fill out the checklist as required as well

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubExportDevices.cs:
```diff
@@ -0,0 +1,73 @@
+    using ResourceProperties = Microsoft.Azure.Commands.Management.IotHub.Properties;
+
+    [Cmdlet(VerbsCommon.New, "AzureRmIotHubExportDevices", SupportsShouldProcess = true)]
+    [OutputType(typeof(JobResponse))]
```
PSIotHubJobResponse?

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubExportDevices.cs:
```diff
@@ -0,0 +1,73 @@
+            Mandatory = true,
+            HelpMessage = "ExcludeKeys")]
+        [ValidateNotNullOrEmpty]
+        public bool ExcludeKeys { get; set; }
```
Why not make this a switch parameter?

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubImportDevices.cs:
```diff
@@ -0,0 +1,73 @@
+    using ResourceProperties = Microsoft.Azure.Commands.Management.IotHub.Properties;
+
+    [Cmdlet(VerbsCommon.New, "AzureRmIotHubImportDevices", SupportsShouldProcess = true)]
+    [OutputType(typeof(JobResponse))]
```
PSIotHubJobResponse?

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHub.cs:
```diff
@@ -0,0 +1,66 @@
+                    }
+                }
+
+                this.WriteObject("Iot Hub Deleted");
```
Do not write text to the output stream.  In the case of delete, you should return nothing for success, and write an error for failure.  You should also implement a PassThrough parameter, and output True on successful delete if it is selected.

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHubEventHubConsumerGroup.cs:
```diff
@@ -0,0 +1,71 @@
+
+        [Parameter(
+            Position = 2,
+            Mandatory = true,
```
Any chance this will come from the pipeline (A Get call, perhaps?)

rkmanda replied:
Nope

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHubEventHubConsumerGroup.cs:
```diff
@@ -0,0 +1,71 @@
+            Mandatory = true,
+            HelpMessage = "Name of the EventHub ConsumerGroupName")]
+        [ValidateNotNullOrEmpty]
+        public string EventHubConsumerGroupName { get; set; }
```
Any chance this will come from the pipeline (A Get call, perhaps?)

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/RemoveAzureRmIotHubKey.cs:
```diff
@@ -0,0 +1,80 @@
+            Position = 2,
+            Mandatory = true,
+            HelpMessage = "Name of the Key")]
+        [ValidateNotNullOrEmpty]
```
Any chance this will come from the pipeline (A Get call, perhaps?)

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs:
```diff
@@ -0,0 +1,248 @@
+    using Microsoft.Azure.Management.IotHub;
+    using Microsoft.Azure.Management.IotHub.Models;
+
+    [Cmdlet(VerbsCommon.Set, "AzureRmIotHub", SupportsShouldProcess = true)]
```
Need a default parameter set setting

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs:
```diff
@@ -0,0 +1,248 @@
+        const string UpdateCloudToDevicePropertiesParameterSet = "UpdateCloudToDeviceProperties";
+        const string UpdateOperationsMonitoringPropertiesParameterSet = "UpdateOperationsMonitoringProperties";
+
+        [Parameter(
```
If this is mandatory in all parameter sets, you can use a single parameter attribute with no ParameterSetName setting

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs:
```diff
@@ -0,0 +1,248 @@
+        [ValidateNotNullOrEmpty]
+        public string ResourceGroupName { get; set; }
+
+        [Parameter(
```
Same comment

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/SetAzureRmIotHub.cs:
```diff
@@ -0,0 +1,248 @@
+
+        [Parameter(
+            ParameterSetName = UpdateFileUploadPropertiesParameterSet,
+            Mandatory = true,
```
Any chance of any of these file settings coming from the pipeline?

rkmanda replied:
Nope

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/ResourceManagerStartup.ps1:
```diff
@@ -0,0 +1,276 @@
+# ----------------------------------------------------------------------------------
```
What is this file intended for?  Remove

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/help/Add-AzureRmIotHubEventHubConsumerGroup.md:
```diff
@@ -0,0 +1,146 @@
+## PARAMETERS
+
+### -EventHubConsumerGroupName
+EventHubConsumerGroupName.
```
Informative description, please

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/help/Set-AzureRmIotHub.md:
```diff
@@ -0,0 +1,307 @@
+
+### -CloudToDevice
+CloudToDevice
+
```
Need an informative description for each parameter.  This applies to all help files

rkmanda commented on src/ResourceManager/IotHub/Commands.IotHub.Test/ScenarioTests/IotHubTests.ps1:
```diff
@@ -0,0 +1,141 @@
+	$op= $iotHubUpdated.Properties.OperationsMonitoringProperties
+	$op.OperationMonitoringEvents["Connections"] = "Information"
+	$iotHubUpdated = Set-AzureRmIotHub -ResourceGroupName $ResourceGroupName -Name $IotHubName -OperationsMonitoringProperties $op
+	Assert-True { $iothubUpdated.Properties.OperationsMonitoringProperties.OperationMonitoringEvents["Connections"] -eq "Information" }
```
Not as a part of the current PR. The test for this is a little involved as it requires a SAS token to be fetched on the fly. Will add the test later.

rkmanda commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/GetAzureRmIotHubJob.cs:
```diff
@@ -0,0 +1,66 @@
+
+    [Cmdlet(VerbsCommon.Get, "AzureRmIotHubJob")]
+    [OutputType(typeof(PSIotHubJobResponse), typeof(List<PSIotHubJobResponse>))]
+    public class GetAzureRmIotHubJob : IotHubBaseCmdlet
```
The prereq for testing this is to have a test for NewAzureRmIotHubExportDevices  which also relies on a valid SAS token to be generated on the fly. Will target this later.

rkmanda commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubExportDevices.cs:
```diff
@@ -0,0 +1,73 @@
+
+    [Cmdlet(VerbsCommon.New, "AzureRmIotHubExportDevices", SupportsShouldProcess = true)]
+    [OutputType(typeof(JobResponse))]
+    public class NewAzureRmIotHubExportDevices : IotHubBaseCmdlet
```
Relies on a valid SAS token to be generated on the fly. Will target this later.

rkmanda commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/NewAzureRmIotHubImportDevices.cs:
```diff
@@ -0,0 +1,73 @@
+
+    [Cmdlet(VerbsCommon.New, "AzureRmIotHubImportDevices", SupportsShouldProcess = true)]
+    [OutputType(typeof(JobResponse))]
+    public class NewAzureRmIotHubImportDevices : IotHubBaseCmdlet
```
Relies on a valid SAS token to be generated on the fly. Will target this later.

rkmanda commented on src/ResourceManager/IotHub/Commands.IotHub/IotHub/GetAzureRmIotHub.cs:
```diff
@@ -0,0 +1,79 @@
+    using Microsoft.Azure.Management.IotHub.Models;
+    using Microsoft.Rest.Azure;
+
+    [Cmdlet(VerbsCommon.Get, "AzureRmIotHub", DefaultParameterSetName = "ListIotHubsBySubscription")]
```
left by mistake

rkmanda commented on src/ResourceManager/IotHub/Commands.IotHub.Test/ScenarioTests/IotHubTests.cs:
```diff
@@ -0,0 +1,43 @@
+        public void TestAzureIotHubLifeCycle()
+        {
+            IotHubController.NewInstance.RunPsTestWorkflow(
+                () => { return new[] { string.Format("{0} {1} {2} {3} {4}", "Test-AzureRmIotHubLifecycle", "northeurope", "powershelliothub", "powershellrg", "S1") }; },
```
Found other resources hardcoding location directly in the script. What I have done is to ensure that the script itself does not have the hardcoding so it can be run in any location. Let me know if this works for the current PR. I will raise an issue to fix this in a subsequent PR.

rkmanda commented:
@markcowl please review. Addressed most comments. Some tests are to be added and I will raise an issue for this. The region hardcoding will also be removed as a part of the test issue.

markcowl commented on src/ResourceManager/IotHub/ChangeLog.md:
```diff
@@ -0,0 +1,23 @@
+## Current Release
+
+## Version 1.0.0
+* Adds commandlets for the Azure IoT Hub 
```
Please listy the capabilities of these cmdlets, and the resources they allow you to manage.

markcowl commented on src/ResourceManager/IotHub/Commands.IotHub/Commands.IotHub.csproj:
```diff
@@ -0,0 +1,209 @@
+    <OutputPath>..\..\..\Package\Release\ResourceManager\AzureResourceManager\AzureRM.IotHub</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
```
Why doi you have two property groups for the release configuration?  Please clean up.

markcowl commented:
on demand run here:http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1339/

signing run here: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-sign/746/

cormacpayne commented:
Signing run: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-sign/753/

cormacpayne commented:
Sign job was successful: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-sign/755/

cormacpayne commented:
On demand: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1346/

Sign job: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-sign/760/

LGTM once both pass

rkmanda commented:
Can this be merged now?
